jax/experimental/shard_map.py

```diff
@@ -0,0 +1,912 @@
+    except ValueError:
+      e, *_ = prefix_errors(in_specs, args)
+      raise e('shard_map in_specs') from None
+    _check_specs_vs_args(f, mesh, in_tree, in_specs, in_specs_flat, args_flat)
+    in_names_flat = tuple(map(_canonicalize_spec, in_specs_flat))
+    flat_fun, out_tree = flatten_fun_nokwargs(fun, in_tree)
+
+    @memoize
+    def out_names_thunk():
+      dummy = tree_unflatten(out_tree(), [object()] * out_tree().num_leaves)
+      try: out_specs_flat = broadcast_prefix(out_specs, dummy)
+      except ValueError:
+        e, *_ = prefix_errors(out_specs, dummy)
+        raise e('shard_map out_specs') from None
+      return tuple(map(_canonicalize_spec, out_specs_flat))
+    try:
+      out_flat = shard_map_p.bind(
+          flat_fun, *args_flat, mesh=mesh, in_names=in_names_flat,
+          out_names_thunk=out_names_thunk, check_rep=check_rep)
+    except _SpecError as e:
+      fails, = e.args
+      msg = _spec_rank_error(SpecErrorType.out, f, out_tree(), out_specs, fails)
+      if any(fail is not no_fail and not fail.shape for fail in fails):
+        msg += (" In particular, for rank 0 outputs which are not constant "
+                "over the mesh, add at least one (singleton) axis to them so "
+                "that they can be concatenated using out_specs.")
+      raise ValueError(msg) from None
+    except _RepError as e:
+      fails, = e.args
+      msg = _rep_error(f, mesh, out_tree(), out_specs, fails)
+      raise ValueError(msg) from None
+    return tree_unflatten(out_tree(), out_flat)
+  return wrapped
+
+# Internally use AxisNames = Dict[int, Tuple[AxisName, ...]], not PartitionSpecs
+AxisName = Hashable
+AxisNames = Dict[int, Tuple[AxisName, ...]]  # TODO(mattjj): make it hashable
+def _canonicalize_spec(spec: PartitionSpec) -> AxisNames:
+  if isinstance(spec, PartitionSpec):
+    return {i: names if isinstance(names, tuple) else (names,)
+            for i, names in enumerate(spec) if names is not None}
+  else:
+    return spec
+
+# Error checking and messages
+
+SpecErrorType = enum.Enum('SpecErrorType', ['input', 'out'])
+
+def _check_specs(error_type: SpecErrorType, specs: Any) -> None:
+  if all(isinstance(p, P) for p in tree_leaves(specs)): return
+  prefix = 'in' if error_type == SpecErrorType.input else 'out'
+  msgs = [f"  {prefix}_specs{key.pprint()} is {x} of type {type(x).__name__}, "
+          for key, x in _generate_key_paths(specs) if not isinstance(x, P)]
+  raise TypeError(
+      f"shard_map {prefix}_specs argument must be a pytree of "
+      f"`jax.sharding.PartitionSpec` instances, but:\n\n"
+      + '\n\n'.join(msgs) + '\n\n'
+      f"Check the {prefix}_specs values passed to shard_map.")
+
+class NoFail: pass
+no_fail = NoFail()
+
+def _check_specs_vs_args(
+    f: Callable, mesh: Mesh, in_tree: PyTreeDef, in_specs: Specs,
+    in_specs_flat: List[P], xs: List) -> None:
+  in_avals = [core.raise_to_shaped(core.get_aval(x)) for x in xs]
```

Comment:
Use api_util.shaped_abstractify? That might be faster here!

Reply:
Done!